### ISSUE TITLE
Failing armhf builds on Docker Hub, latest-armhf tag 2 months old

### ISSUE THREAD

**Issue by chilicheech** [haugene/docker-transmission-openvpn#721]
Is the latest-armhf not being built anymore? Did something change?

Also, why do we build the armhf images off of `resin/rpi-raspbian:stretch` instead of `ubuntu` like the non-armhf image? Shouldn't we use the `ubuntu` image to keep it consistent?

**Comment by haugene** [comment]
The builds are still running, but they're failing and so the tag isn't updated. This is due to #705.

As for your second question, you mean some armhf version of ubuntu? Definitely open to that if you got one in mind. But we can't base it of the main ubuntu image. Described here for example: https://blog.hypriot.com/getting-started-with-docker-on-your-arm-device/
> Docker-based apps you use have to be packaged specifically for ARM architecture! Docker-based apps packaged for x86/x64 will not work and will result in an error

If we could base it off the main ubuntu there wouldn't be an armhf version of the image.

**Comment by chilicheech** [comment]
There is an armhf based ubuntu image. I use it to build images for my raspberry pi all the time. It isn't tagged as docker now has an automated way to determine which image to pull based on the target architecture. For example:
- when I run `docker pull ubuntu` on a Mac or an x86 machine it pulls the x86 ubuntu image
- when I run `docker pull ubuntu` on a raspberry pi or an armhf machine it pulls the armhf ubuntu image

In a Dockerfile I simply have to put `FROM ubuntu` and it will know to pull the correct architecture based on the target architecture it's building the image for. From what I understand we can just use the original Dockerfile to build the image for both x86 and armhf. I'm not sure on how to get an automated build to work like that but I should be able to build the Dockerfile on my raspberry pi.

I tried building the Dockerfile on my raspberry pi and got an issue with the rar package not being in any repo, and some other dependency issues. But in general, it should be possible to use the same Dockerfile for both x86 and arm as long as the dependencies are available.

I realize this type of work might take time and might not be a priority right now...

**Comment by haugene** [comment]
Aha, cool. Didn't know that. We're leaning on the Docker Hub automated builds and they're only building on x86/x64 architecture as far as I know. That's why we need the cross-build command, and a separate Dockerfile for it.

I guess there are many potential improvements to how it's done today. Time and priority is what it comes down to as you say. I haven't really developed or maintained the armhf image at all, just accepted pull-requests and hoped for the best. I don't have a raspberry at the moment and can't really verify the changes myself.

There are many smaller issues that I would like to fix and give the project a good "brush up", but then there's finding time. I follow up issues and merge PRs on my commute, and that's about it these days.

Anyways. One day :) I think I'll close this now, if that's ok with you? Feel free to create another issue as a proposed enhancement if you have a suggestion on how to do it.

**Comment by haugene** [comment]
There are still issues with the automated builds, even after #723. Since you say that it builds correctly on arm devices then this is probably just the cross-build stuff failing for some reason. Might not be included in the new base image, and the question is then - how do we add it again? :)

**Comment by haugene** [comment]
Ok, I'm ready to fail - but did some research and will be changing the base image again.
Based on [this list](https://www.balena.io/docs/reference/base-images/base-images-ref/) of base images I've chosen to go with `balenalib/raspberrypi3:stretch`

I'm now successfully building the armhf image on my mac, and it should work on the Docker hub as well. Note that even if it says raspberrypi3, the only difference to the raspberrypi2 image is a label. See the corresponding Dockerfiles here:
[RaspberryPi2](https://github.com/balena-io-library/base-images/blob/master/balena-base-images/device-base/raspberry-pi2/debian/stretch/run/Dockerfile)
[RaspberryPi3](https://github.com/balena-io-library/base-images/blob/master/balena-base-images/device-base/raspberrypi3/debian/stretch/run/Dockerfile)

The two images also share a parent image. Was considering using it as well but it seems that the updates to sources list might come in handy. Hopefully this will fix it.

**Comment by haugene** [comment]
Cleaning up issues and closing this for now. Not merged to master yet, but that should be done soon. Maybe a weeks time, trying to do some more stuff at the same time 😄

**Comment by kushagharahi** [comment]
Can I ask why you choose to go with `balenalib/raspberrypi3:stretch` rather than `balenalib/raspberry-pi` as the arm base image? I personally run your fantastic docker image on a pi zero and the pi3 image does not build/run on a pi zero. I do believe that the raspberry-pi image will ensure that your image works on rpi1/2/3/Zero. If you would like, I can PR `balenalib/raspberry-pi` as the base image into dev.

Thanks very much for your project!

**Comment by haugene** [comment]
Hey @kushagharahi, I guess I just got lost in the jungle of balenalib's base images.
I chose it because I thought it was the best match, but I'm happy to get some help and be corrected 😄 

Just committed a change of base image for all armhf Dockerfiles. Glad to hear you like the project, and you're welcome :)

**Comment by kushagharahi** [comment]
You rock, thank you very much
